sklearn/metrics/_dist_metrics.pyx.tp

```diff
@@ -660,83 +678,93 @@ cdef class DistanceMetric{{name_suffix}}:
 
             Darr = np.empty((n_X, n_Y), dtype=DTYPE, order='C')
             self.cdist_csr(
-                x1_data=X_data,
+                x1_data=&X_data[0],
                 x1_indices=X_indices,
                 x1_indptr=X_indptr,
-                x2_data=Y_data,
+                x2_data=&Y_data[0],
                 x2_indices=Y_indices,
                 x2_indptr=Y_indptr,
-                size=size,
+                size=n_features,
                 D=Darr,
             )
-        return Darr
-
-    def _pairwise_sparse_dense(self, X, Y):
-        n_X, size = X.shape
-        X_data = np.asarray(X.data, dtype={{INPUT_DTYPE}})
-        X_indices = np.asarray(X.indices, dtype=SPARSE_INDEX_TYPE)
-        X_indptr = np.asarray(X.indptr, dtype=SPARSE_INDEX_TYPE)
-
-        # To avoid introducing redundant implementations for the CSR × dense array
-        # case, we wrap the dense array into a fake CSR datastructure and leverage
-        # the existing code for the CSR × CSR case.
-        # The true CSR representation of a dense array would require allocating
-        # a Y_indices matrix of shape (n_samples, n_features) with repeated
-        # contiguous integers from 0 to n_features - 1 on each row which would
-        # be very wasteful from a memory point of view. Instead we only allocate
-        # a single row and adapt the CSR × CSR routines to use a modulo operation
-        # when accessing Y_indices in order to achieve the same result without having
-        # to materialize the indices repetition explicitly.
-
-        n_Y, _ = Y.shape
-        Y_data = Y.reshape(-1)
-        Y_indices = np.arange(size, dtype=SPARSE_INDEX_TYPE)
-        Y_indptr = np.arange(
-            start=0, stop=size * (n_Y + 1), step=size, dtype=SPARSE_INDEX_TYPE
-        )
+        return np.asarray(Darr)
+
+    def _pairwise_sparse_dense(self,
+        X: csr_matrix,
+        const {{INPUT_DTYPE_t}}[:, ::1] Y,
+    ):
+        cdef:
+            ITYPE_t n_X = X.shape[0]
+            ITYPE_t n_features = X.shape[1]
+            const {{INPUT_DTYPE_t}}[:] X_data = np.asarray(X.data, dtype={{INPUT_DTYPE}})
+            const SPARSE_INDEX_TYPE_t[:] X_indices = np.asarray(X.indices, dtype=SPARSE_INDEX_TYPE)
+            const SPARSE_INDEX_TYPE_t[:] X_indptr = np.asarray(X.indptr, dtype=SPARSE_INDEX_TYPE)
+
+            ITYPE_t n_Y = Y.shape[0]
+            const {{INPUT_DTYPE_t}}[:] Y_data = np.reshape(Y, -1)
+
+            # Potentially costly to store, but allows using `cdist_csr`
+            # without logic duplication.
+            const SPARSE_INDEX_TYPE_t[:] Y_indices = np.tile(
+                np.arange(n_features, dtype=SPARSE_INDEX_TYPE), reps=n_Y,
+            )
```

Comment:
Can you please try again origin 2 solutions explored in https://github.com/jjerphan/scikit-learn/pull/15/files that did not involved materializing the `Y_indices` with `np.tile` but with the fix to `Y_indptr` below?

If we decide to keep on using `np.tile` for some reason, then we could keep the memoryviews for `x1_data` and `x2_data`, right?

Reply:
but I would really like to avoid the extra memory allocation of `np.tile` if it can be avoided.

Reply:
Done via 732b464ec4ff6a35e1e12391512496ec9b08ae69.